subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -62,38 +82,63 @@ public void changed(Collection<CompleteFileSystemLocationSnapshot> removedSnapsh
         updateWatchedDirectories();
     }
 
+    @Override
+    public void buildFinished() {
+        recentlyWatchedProjectRootDirectories.clear();
+        oldProjectRootDirectories.putAll(projectRootDirectories);
+        projectRootDirectories.clear();
+        updateWatchedDirectories();
+        oldProjectRootDirectories.keySet().retainAll(watchedRoots);
+    }
+
     @Override
     public void updateProjectRootDirectories(Collection<File> updatedProjectRootDirectories) {
         Set<Path> rootPaths = updatedProjectRootDirectories.stream()
             .map(File::toPath)
             .map(Path::toAbsolutePath)
             .collect(Collectors.toSet());
+        Set<Path> newProjectRootDirectories = WatchRootUtil.resolveRootsToWatch(rootPaths);
+        LOGGER.info("Now considering {} as root directories to watch", newProjectRootDirectories);
+
+        oldProjectRootDirectories.keySet().removeAll(newProjectRootDirectories);
+
         projectRootDirectories.clear();
-        for (Path rootPathToWatch : WatchRootUtil.resolveRootsToWatch(rootPaths)) {
-            projectRootDirectories.put(rootPathToWatch, rootPathToWatch.toString() + File.separator);
-        }
-        LOGGER.info("Now considering {} as root directories to watch", projectRootDirectories.keySet());
+        newProjectRootDirectories.forEach(
+            oldProjectRootDirectory -> projectRootDirectories.put(oldProjectRootDirectory, oldProjectRootDirectory.toString() + File.separator)
+        );
+
```

Comment:
There is just way too many sets of root directories here, old, new, un-prefixed, recently watched and watched. Shall we do a quick pairing session to see if we can find a way to make it easier to understand?